diff --git a/src/Quarks.c b/src/Quarks.c
index 0dd91968..dbaf5086 100644
--- a/src/Quarks.c
+++ b/src/Quarks.c
@@ -105,6 +105,11 @@ static XrmQuark nextUniq = -1;	/* next quark from XrmUniqueQuark */
 #define CLEARPERM(q) BYTEREF(q) &= ~(1 << ((q) & 7))
 #endif
 
+#undef _XLockMutex
+#undef _XUnlockMutex
+#define _XLockMutex(m)
+#define _XUnlockMutex(m)
+
 /* Permanent memory allocation */
 
 #define WALIGN sizeof(unsigned long)
diff --git a/src/Xrm.c b/src/Xrm.c
index a8ebf883..d716fdb0 100644
--- a/src/Xrm.c
+++ b/src/Xrm.c
@@ -301,6 +301,13 @@ typedef unsigned char XrmBits;
 #define is_simple(bits)		((bits) & (NORMAL|BSLASH))
 #define is_special(bits)	((bits) & (ENDOF|BSLASH))
 
+#undef _XLockMutex
+#undef _XUnlockMutex
+#undef _XCreateMutex
+#define _XLockMutex(m)
+#define _XUnlockMutex(m)
+#define _XCreateMutex(m)
+
 /* parsing types */
 static XrmBits const xrmtypes[256] = {
     EOS,0,0,0,0,0,0,0,
@@ -500,9 +507,7 @@ static XrmDatabase NewDatabase(void)
 	_XCreateMutex(&db->linfo);
 	db->table = (NTable)NULL;
 	db->mbstate = (XPointer)NULL;
-	db->methods = _XrmInitParseInfo(&db->mbstate);
-	if (!db->methods)
-	    db->methods = &mb_methods;
+	db->methods = &mb_methods;
     }
     return db;
 }
